lib/typecheck.ts

```diff
@@ -368,6 +383,23 @@ export default class TypeChecker {
             throw new TypeError(`Invalid principal ${principal}, must be a contact or a group`);
     }
 
+    private _expandTypeScope(typeScope : TypeScope, key : string) : TypeScope[] {
+        if (!(key in typeScope))
+            return [typeScope];
+        if (!(typeScope[key] instanceof EntityType))
+            return [typeScope];
+        
+        const entityType = (typeScope[key] as EntityType).type;
```

Comment:
Do you need this type assertion? You did the instanceof check right before it, doesn't the compiler pick it up?
If it doesn't, maybe with a temporary variable?